File: cpp/bin/sqlt.cpp
// The SEL->SQL conformance suite for the C++ host.
//
// Run from the repository root:
//
//     cpp/build/sqlt                 every case
//     cpp/build/sqlt bind. agg.      only cases whose name contains one of these
//     cpp/build/sqlt --names         what this host loaded, and stop
//
// The cases live in sql/cases/*.sqlt and reach here through
// tools/gen-sql-cases.mjs, which is the only thing that reads them. Nothing in
// this file parses anything: a case's bindings and registrations arrive as
// typed constructor calls the compiler has already checked.

#include <cstdio>
#include <cstdlib>
#include <map>
#include <set>
#include <stdexcept>
#include <string>
#include <vector>

#include "../sel.hpp"
#include "../sel_sql.hpp"
#include "case_data.hpp"

namespace {

using namespace sel::sqlt;
using sel::sql::Fragment;
using sel::sql::Map;
using sel::sql::Mode;
using sel::sql::Sql;
using sel::sql::SqlError;

// A case file names PHP's exception class, because the cases were written for
// one host. The line it draws is the one that matters and it is the same
// everywhere: a malformed map or registration is a mistake in the application's
// startup, not a rule that cannot be translated, so it must NOT be catchable as
// SqlError -- try_translate() swallows the second and must not swallow the
// first. A name with no mapping is a suite error and never a pass: an
// unrecognised expectation must not be satisfied by whatever happened to be
// thrown.
bool throws_is_known(const std::string& name) { return name == "LogicException"; }

const std::map<std::string, std::string> MIRRORS = {{"mariadb", "mysql"}};

struct SuiteError : std::runtime_error {
  using std::runtime_error::runtime_error;
};

// "CODE" or "CODE line:col".
struct Expected {
  std::string code;
  bool has_pos = false;
  int line = 0, col = 0;
};

Expected parse_expected(const std::string& s, const std::string& at) {
  Expected e;
  const std::size_t sp = s.find(' ');
  if (sp == std::string::npos) { e.code = s; return e; }
  e.code = s.substr(0, sp);
  std::string pos = s.substr(sp + 1);
  while (!pos.empty() && pos.front() == ' ') pos.erase(pos.begin());
  const std::size_t colon = pos.find(':');
  if (colon == std::string::npos) throw SuiteError(at + ": malformed error expectation");
  try {
    e.line = std::stoi(pos.substr(0, colon));
    e.col = std::stoi(pos.substr(colon + 1));
  } catch (const std::exception&) {
    throw SuiteError(at + ": malformed error expectation");
  }
  e.has_pos = true;
  return e;
}

// `~1~`, not bare tildes: PostgreSQL's regex operator IS `~`, so counting those
// divided a regex fragment's odd tilde count by two. PHP's runner shipped that
// bug; every host has carried the corrected form since.
int count_slots(const std::string& s) {
  int n = 0;
  for (std::size_t i = 0; i + 2 < s.size(); ++i) {
    if (s[i] != '~') continue;
    std::size_t j = i + 1;
    while (j < s.size() && s[j] >= '0' && s[j] <= '9') ++j;
    if (j > i + 1 && j < s.size() && s[j] == '~') { ++n; i = j; }
  }
  return n;
}

std::string join_dumps(const std::vector<sel::Value>& vs) {
  std::string out;
  for (std::size_t i = 0; i < vs.size(); ++i) {
    if (i) out += ", ";
    out += vs[i].dump();
  }
  return out;
}

// Empty when the case passes, else what went wrong.
std::string run_case(const SqlCase& c, const std::string& dialect) {
  if (dialect.empty()) {
    throw SuiteError(std::string(c.at) + ": case " + c.name + " has no --- dialect");
  }
  const std::string as_ = c.as_ ? c.as_ : "value";
  const std::string mode_name = c.mode ? c.mode : "inline";
  const auto mode = sel::sql::mode_from_name(mode_name);
  if (!mode) throw SuiteError(std::string(c.at) + ": unknown mode " + mode_name);

  bool have_sql = false, have_error = false, have_thrown = false;
  std::string sql, thrown_what;
  SqlError error("", "");
  Fragment frag;

  try {
    // Inside the try: a bad registration is one of the outcomes a case may
    // assert, so it has to be catchable rather than fatal.
    if (c.register_fn) c.register_fn();
    // Inside the try: a Binding constructor refuses a malformed binding at the
    // earliest possible moment, which is construction rather than translation,
    // and that refusal is one of the outcomes a case asserts.
    sel::sql::Bindings bindings(c.bindings_fn());
    sel::Program program = sel::compile(c.source);
    sel::sql::Options options;
    options.strict = c.strict;
    frag = Sql::translate(program, dialect, bindings, options);
    sql = as_ == "condition" ? frag.as_condition(*mode) : frag.as_value(*mode);
    have_sql = true;
  } catch (const SqlError& e) {
    error = e;
    have_error = true;
  } catch (const sel::SelError& e) {
    return "the source did not compile: " + e.str();
  } catch (const SuiteError&) {
    throw;
  } catch (const std::exception& e) {
    thrown_what = e.what();
    have_thrown = true;
  }

  if (c.throws) {
    if (!throws_is_known(c.throws)) {
      throw SuiteError(std::string(c.at) + ": no C++ equivalent is recorded for "
                       "--- throws " + c.throws);
    }
    if (!have_thrown) {
      return std::string("expected ") + c.throws + ", got " +
             (have_error ? error.str() : sql);
    }
    return "";
  }
  if (have_thrown) {
    throw SuiteError(std::string(c.at) + ": unexpected throw: " + thrown_what);
  }

  if (c.error) {
    if (!have_error) return std::string("expected ") + c.error + ", got " + sql;
    const Expected want = parse_expected(c.error, c.at);
    if (error.code() != want.code) {
      return "expected " + want.code + ", got " + error.code() + " (" +
             error.message() + ")";
    }
    if (want.has_pos) {
      const std::string got = std::to_string(error.line()) + ":" +
                              std::to_string(error.col());
      const std::string wanted = std::to_string(want.line) + ":" +
                                 std::to_string(want.col);
      if (got != wanted) {
        return "expected " + want.code + " at " + wanted + ", got it at " + got;
      }
    }
    return "";
  }

  if (have_error) {
    return "expected SQL, got " + error.code() + " (" + error.message() + ")";
  }
  if (!have_sql || sql != std::string(c.expect ? c.expect : "")) {
    return "got:  " + sql + "\n     want: " + std::string(c.expect ? c.expect : "");
  }

  // Checked for every case that produces a fragment, not only those asking
  // about params: every slot in the part list must have a value, and every value
  // must be emitted. A value bound but never emitted means a Fragment was
  // rendered and thrown away -- invisible in inline mode, which is what the rest
  // of the suite asserts.
  std::set<int> seen;
  for (const Fragment::Part& p : frag.parts()) {
    if (!p.is_slot) continue;
    if (p.slot < 1 || static_cast<std::size_t>(p.slot) > frag.params().size()) {
      return "parameter slot " + std::to_string(p.slot) + " has no value in params";
    }
    seen.insert(p.slot);
  }
  std::string orphans;
  for (std::size_t i = 1; i <= frag.params().size(); ++i) {
    if (!seen.count(static_cast<int>(i))) {
      if (!orphans.empty()) orphans += ", ";
      orphans += std::to_string(i);
    }
  }
  if (!orphans.empty()) {
    return "parameter slot(s) [" + orphans +
           "] were bound but never emitted — a fragment was rendered and discarded";
  }
  if (static_cast<int>(frag.bindings().size()) !=
      count_slots(frag.as_value(Mode::Debug))) {
    return "bindings() and the emitted placeholders disagree in count";
  }

  if (c.params) {
    const std::string got = join_dumps(frag.bindings());
    if (got != c.params) {
      return "params got:  " + got + "\n     want: " + c.params;
    }
  }
  return "";
}

}  // namespace

int main(int argc, char** argv) {
  std::vector<std::string> filters(argv + 1, argv + argc);
  const auto cases = sql_cases();

  if (filters.size() == 1 && filters[0] == "--names") {
    for (const SqlCase& c : cases) std::printf("%s\t%s\n", c.at, c.name);
    return 0;
  }

  int passed = 0, mirrored = 0, compile_refused = 0, suite_errors = 0;
  std::vector<std::pair<const SqlCase*, std::string>> failures;

  for (const SqlCase& c : cases) {
    if (!filters.empty()) {
      bool wanted = false;
      for (const std::string& f : filters) {
        if (std::string(c.name).find(f) != std::string::npos) { wanted = true; break; }
      }
      if (!wanted) continue;
    }

    // The case asserts that this shape is refused, and here it was refused by
    // the compiler: the typed constructors cannot express it at all. Counted
    // and named rather than skipped, so the total stays honest.
    if (c.unrepresentable) {
      ++passed;
      ++compile_refused;
      continue;
    }

    Map::reset();   // no case may leak a registration into another
    std::string problem;
    try {
      problem = run_case(c, c.dialect);
    } catch (const SuiteError& e) {
      std::printf("SUITE ERROR %s\n", e.what());
      ++suite_errors;
      continue;
    }
    if (!problem.empty()) { failures.emplace_back(&c, problem); continue; }
    ++passed;

    // The same case under the mirrored dialect. A registration case is exempt:
    // it names its dialect in the register data, so re-running it under another
    // name would be testing something it does not claim.
    auto m = MIRRORS.find(c.dialect);
    if (m == MIRRORS.end() || c.register_fn) continue;
    Map::reset();
    try {
      problem = run_case(c, m->second);
    } catch (const SuiteError& e) {
      std::printf("SUITE ERROR (mirrored to %s) %s\n", m->second.c_str(), e.what());
      ++suite_errors;
      continue;
    }
    if (problem.empty()) ++mirrored;
    else {
      failures.emplace_back(&c, "mirrored to " + m->second +
                                    ", which must agree with " + c.dialect + ": " +
                                    problem);
    }
  }

  for (const auto& [c, problem] : failures) {
    std::printf("FAIL %s  (%s)\n     %s\n", c->name, c->at, problem.c_str());
  }
  std::printf("\n%d passed (%d also checked against a mirrored dialect, "
              "%d refused by the type system), %d failed, %d suite errors\n",
              passed, mirrored, compile_refused,
              static_cast<int>(failures.size()), suite_errors);
  return failures.empty() && suite_errors == 0 ? 0 : 1;
}

File: js/bin/sqlt.mjs
#!/usr/bin/env node
// Run sql/cases/*.sqlt against the JS translator.
//
// The counterpart of php/bin/sqlt, and deliberately not all of it, on the same
// split python/bin/sqlt draws. That runner has five checks; three ask about the
// HOST and are here, two ask about the MAP and the SUITE and are not:
//
//   here   runCase           the emitted string, the error code and position, the
//                            parameter list — what this host does with shared data
//   here   the mirror        re-run every mariadb case as mysql, requiring the same
//                            string: does THIS host emit identical SQL for a leaf
//                            that overrides nothing
//   here   the slot checks   every bound value is emitted, and bindings() agrees in
//                            count with the placeholders
//
//   not    caveat pins       every caveated entry is pinned by some case. A property
//                            of the case files against the dialect data, both shared
//                            and both already checked by PHP.
//   not    mirror parity     mariadb and mysql declare the same map. Likewise a
//                            property of sql/dialects/*.json and nothing to do with
//                            which host reads it.
//
// That split is why js/src/sql/map.mjs has no trace facility: nothing here needs
// one, and adding one would only be to re-run a check PHP already owns.
//
//     node js/bin/sqlt.mjs [name-substring ...]

import { compile } from '../src/sel.mjs';
import { SelError } from '../src/errors.mjs';
import { Sql, SqlError, map as sqlmap } from '../src/sql/index.mjs';
import { optimizeAstInMemory } from '../src/optimizer.mjs';
import { SQL_CASES } from './case-data.mjs';

// A malformed suite. Not a failing case — a suite that cannot be run.
class SuiteError extends Error {}

const MIRRORS = { mariadb: 'mysql' };

// `--- throws` names PHP's class, because the cases were written for one host.
// The line it draws is the one that matters and it is the same in every host: a
// malformed map or registration is a mistake in the application's startup, not a
// rule that cannot be translated, so it must NOT be catchable as SqlError —
// tryTranslate() swallows the second and must not swallow the first. PHP spells
// that LogicException, Python RuntimeError; this host spells it a plain Error,
// which is what js/src/registry.mjs already throws for a function defined twice.
//
// Matched on the exact constructor rather than instanceof: every host error is an
// Error here, so `instanceof Error` would accept a TypeError from a genuine bug
// as though the case had asserted it.
//
// A name with no entry here is a suite error rather than a pass: an expectation
// nobody has mapped must never be satisfied by whatever happened to be thrown.
const THROWS = { LogicException: Error };

// A case's runtime registrations, applied before translating.
function applyRegistrations(ops) {
  if (ops === null || ops === undefined) return;
  for (const op of ops) {
    if (Object.hasOwn(op, 'define')) {
      const [dialect, section, key, entry] = op.define;
      sqlmap.define(dialect, section, key, entry);
      continue;
    }
    if (Object.hasOwn(op, 'dialect')) {
      const rest = {};
      for (const [k, v] of Object.entries(op)) if (k !== 'dialect') rest[k] = v;
      sqlmap.defineDialect(op.dialect, rest);
      continue;
    }
    throw new SuiteError('a register op needs a dialect or a define');
  }
}

const ERROR_RE = /^(\S+)(?:\s+(\d+):(\d+))?$/;
const TILDE_RE = /~\d+~/g;

// The tree as text, with the registry's Spec objects left out: they are looked
// up by name and compared by identity, and a snapshot is compared by value.
// Everything else that identifies a node -- kind, position, literal, name,
// operator, grouping, children -- is in here, so two snapshots are equal
// exactly when the caller would see the same tree.
function snapshotAst(ast) {
  return JSON.stringify(ast, (key, value) =>
    key === 'spec' ? undefined : typeof value === 'bigint' ? value.toString() : value);
}

// Report what went wrong with a planner case, or null.
//
// A `--- plan` case asks the planner rather than the translator. It asserts the
// classification, the physical sources, the SQL prefix, that the continuation
// exists exactly when the classification says so, and that the caller's AST is
// the same tree afterwards -- after planning, which runs stage 1 and the logical
// optimiser, and after the physical optimiser Program.run() uses.
function runPlanCase(c) {
  const dialect = c.dialect;
  const options = c.options || {};
  const mode = c.mode || 'inline';
  let plan = null;
  let error = null;
  let program = null;
  let before = null;
  try {
    applyRegistrations(c.register);
    const bindings = c.bindings();
    program = compile(c.source);
    before = snapshotAst(program.ast);
    plan = Sql.planHybrid(program, dialect, bindings, options);
  } catch (e) {
    if (e instanceof SuiteError) throw e;
    else if (e instanceof SqlError) error = e;
    else if (e instanceof SelError) return `the source did not compile: ${e}`;
    else throw new SuiteError(`${c.at}: unexpected ${e.constructor.name}: ${e.stack}`);
  }

  if (c.plan === 'refused') {
    if (error === null) return `expected ${c.error}, got a ${classify(plan)} plan`;
    const m = ERROR_RE.exec(c.error);
    if (m === null) throw new SuiteError(`${c.at}: malformed error expectation`);
    if (error.code !== m[1]) return `expected ${m[1]}, got ${error.code} (${error.message})`;
    return null;
  }
  if (error !== null) return `expected a ${c.plan} plan, got ${error.code} (${error.message})`;

  const got = classify(plan);
  if (got !== c.plan) return `expected a ${c.plan} plan, got ${got}`;
  if (c.tables !== null && c.tables !== undefined) {
    const want = JSON.stringify(c.tables);
    const have = JSON.stringify(plan.sourceTables);
    if (want !== have) return `source tables got:  ${have}\n     want: ${want}`;
  }
  if (plan.dialect !== dialect) return `plan.dialect is ${plan.dialect}, not ${dialect}`;

  if (c.plan === 'pure_memory') {
    if (plan.sqlStatement !== null) return 'a pure-memory plan carries a SQL statement';
    if (plan.continuationProgram !== program) {
      return 'a pure-memory plan must run the original program';
    }
    if (plan.continuationAst !== program.ast) {
      return 'a pure-memory plan must expose the original AST as its continuation';
    }
  } else {
    if (plan.sqlStatement === null) return `a ${c.plan} plan has no SQL statement`;
    if (plan.sqlPrefixAst === null) return `a ${c.plan} plan has no SQL prefix AST`;
    const sql = plan.sqlStatement.asStatement(mode);
    if (sql !== c.expect) return `got:  ${sql}\n     want: ${c.expect}`;
    if (c.plan === 'pure_sql') {
      if (plan.continuationProgram !== null || plan.continuationAst !== null) {
        return 'a pure-SQL plan carries a continuation';
      }
    } else if (plan.continuationProgram === null || plan.continuationAst === null) {
      return 'a hybrid plan has no continuation';
    }
  }

  // Planning must not have touched the tree, and neither may the physical
  // optimiser that every run() goes through.
  if (snapshotAst(program.ast) !== before) return 'planning mutated the program AST';
  optimizeAstInMemory(program.ast);
  if (snapshotAst(program.ast) !== before) {
    return 'the physical optimiser mutated the program AST';
  }
  return null;
}

function classify(plan) {
  if (plan.pureSql) return 'pure_sql';
  if (plan.pureMemory) return 'pure_memory';
  return 'hybrid';
}

// null when the case passes, else what went wrong.
function runCase(c) {
  const dialect = c.dialect;
  if (dialect === null || dialect === '') {
    throw new SuiteError(`${c.at}: case ${c.name} has no --- dialect`);
  }
  const options = c.options || {};
  const as = c.as || 'value';
  const mode = c.mode || 'inline';

  let sql = null;
  let error = null;
  let thrown = null;
  let frag = null;
  let program = null;
  let bindings = null;
  try {
    // Inside the try: a bad registration is one of the outcomes a case may
    // assert, so it has to be catchable rather than fatal.
    applyRegistrations(c.register);
    // Inside the try: a Binding constructor refuses a malformed binding at the
    // earliest possible moment, which is construction rather than translation,
    // and that refusal is one of the outcomes a case asserts.
    bindings = c.bindings();
    program = compile(c.source);
    frag = Sql.translate(program, dialect, bindings, options);
    sql = as === 'condition' ? frag.asCondition(mode)
      : as === 'statement' ? frag.asStatement(mode)
      : frag.asValue(mode);
  } catch (e) {
    if (e instanceof SuiteError) throw e;
    else if (e instanceof SqlError) error = e;
    else if (e instanceof SelError) return `the source did not compile: ${e}`;
    else thrown = e;
  }

  // A malformed map or binding is a mistake in the application's startup and
  // throws the host's own error type; a rule that cannot be translated throws
  // SqlError. Keeping the two apart is why `throws` exists as its own outcome
  // rather than as another E_SQL_ code.
  if (c.throws !== null && c.throws !== undefined) {
    const want = Object.hasOwn(THROWS, c.throws) ? THROWS[c.throws] : null;
    if (want === null) {
      throw new SuiteError(`${c.at}: no JS equivalent is recorded for --- throws `
        + `${c.throws}; add one to THROWS`);
    }
    if (thrown === null) {
      return `expected ${c.throws}, got ${error ? String(error) : JSON.stringify(sql)}`;
    }
    if (thrown.constructor === want) return null;
    return `expected ${c.throws} (${want.name}), got ${thrown.constructor.name} (${thrown})`;
  }
  if (thrown !== null) {
    throw new SuiteError(`${c.at}: unexpected ${thrown.constructor.name}: ${thrown.stack}`);
  }

  // Every `--- as statement` case is also run through translateStatement, the
  // public full-delegation entry point, which must say exactly what
  // translate() says -- the same text, or the same refusal at the same
  // column. Two hosts ran the logical optimiser in that lane and three did
  // not, and only a twin check can see it (review 2026-09-15 finding C).
  if (as === 'statement' && program !== null) {
    let twinSql = null;
    let twinError = null;
    try {
      twinSql = Sql.translateStatement(program, dialect, bindings, options).asStatement(mode);
    } catch (e) {
      if (e instanceof SqlError) twinError = e;
      else throw new SuiteError(`${c.at}: translateStatement threw ${e.constructor.name}: ${e.stack}`);
    }
    if (error !== null || twinError !== null) {
      const got = (e) => (e === null ? 'SQL' : `${e.code} at ${e.line}:${e.col}`);
      if (error === null || twinError === null || error.code !== twinError.code
          || error.line !== twinError.line || error.col !== twinError.col) {
        return `translate() gave ${got(error)} but translateStatement() gave ${got(twinError)}`;
      }
    } else if (twinSql !== sql) {
      return `translateStatement() disagrees with translate():\n     ${twinSql}\n     ${sql}`;
    }
  }

  if (c.error !== null && c.error !== undefined) {
    if (error === null) return `expected ${c.error}, got ${JSON.stringify(sql)}`;
    const m = ERROR_RE.exec(c.error);
    if (m === null) throw new SuiteError(`${c.at}: malformed error expectation`);
    if (error.code !== m[1]) return `expected ${m[1]}, got ${error.code} (${error.message})`;
    if (m[2] !== undefined) {
      const wantPos = `${m[2]}:${m[3]}`;
      const gotPos = `${error.line}:${error.col}`;
      if (wantPos !== gotPos) return `expected ${m[1]} at ${wantPos}, got it at ${gotPos}`;
    }
    return null;
  }

  if (error !== null) return `expected SQL, got ${error.code} (${error.message})`;
  if (sql !== c.expect) return `got:  ${sql}\n     want: ${c.expect}`;

  // Checked for every case that produces a fragment, not only those asking about
  // params: every slot in the part list must have a value, and every value must
  // be emitted. A value bound but never emitted means a Fragment was rendered and
  // thrown away — invisible in `inline` mode, which is what the rest of the suite
  // asserts.
  const seen = new Set();
  for (const p of frag.parts) {
    if (typeof p === 'string') continue;
    if (p < 1 || p > frag.params.length) {
      return `parameter slot ${p} has no value in params`;
    }
    seen.add(p);
  }
  const orphans = [];
  for (let i = 1; i <= frag.params.length; i += 1) if (!seen.has(i)) orphans.push(i);
  if (orphans.length) {
    return `parameter slot(s) ${JSON.stringify(orphans)} were bound but never `
      + 'emitted — a fragment was rendered and discarded';
  }
  // `~1~`, not bare tildes: PostgreSQL's regex operator IS `~`, so counting them
  // divided a regex fragment's odd tilde count by two. PHP's runner shipped that
  // bug; this is the corrected form from the start.
  const debugStr = frag.kind === 'STATEMENT' ? frag.asStatement('debug') : frag.asValue('debug');
  const tildes = debugStr.match(TILDE_RE) || [];
  if (frag.bindings().length !== tildes.length) {
    return 'bindings() and the emitted placeholders disagree in count';
  }

  if (c.params !== null && c.params !== undefined) {
    const got = frag.bindings().map((v) => v.dump()).join(', ');
    if (got !== c.params) return `params got:  ${got}\n     want: ${c.params}`;
  }
  return null;
}

function main(argv) {
  const filters = argv.slice(2);
  const cases = SQL_CASES;

  // `--names` prints what this host loaded and stops. Kept as a cheap way to see
  // the roster; the hosts cannot disagree about it any more, because
  // tools/gen-sql-cases.mjs is the only thing that reads sql/cases/*.sqlt and
  // they all load what it wrote.
  if (filters.length === 1 && filters[0] === '--names') {
    for (const c of cases) process.stdout.write(`${c.at}\t${c.name}\n`);
    return 0;
  }

  let passed = 0;
  let mirrored = 0;
  const failures = [];
  let suiteErrors = 0;

  for (const c of cases) {
    if (filters.length && !filters.some((f) => c.name.includes(f))) continue;
    sqlmap.reset();        // no case may leak a registration into another
    let problem;
    try {
      problem = c.plan ? runPlanCase(c) : runCase(c);
    } catch (e) {
      if (!(e instanceof SuiteError)) throw e;
      process.stdout.write(`SUITE ERROR ${e.message}\n`);
      suiteErrors += 1;
      continue;
    }
    if (problem !== null) { failures.push([c, problem]); continue; }
    passed += 1;

    // The same case under the mirrored dialect. A registration case is exempt: it
    // names its dialect in the register data, so re-running it under another name
    // would be testing something it does not claim.
    const mirror = Object.hasOwn(MIRRORS, c.dialect) ? MIRRORS[c.dialect] : null;
    if (mirror === null || (c.register !== null && c.register !== undefined)) continue;
    sqlmap.reset();
    // Spread is RIGHT-wins here, where PHP's `+` is left-wins; the assertion
    // below is the same one php/bin/sqlt carries, because getting this backwards
    // re-runs every mariadb case as mariadb and the count line still prints a
    // plausible number.
    const mirrorCase = { ...c, dialect: mirror };
    // Pin the one SQL spelling difference: each server's NO PAD collation.
    if (mirrorCase.expect) mirrorCase.expect = mirrorCase.expect.replaceAll(
      ' COLLATE utf8mb4_nopad_bin', ' COLLATE utf8mb4_0900_bin');
    if (mirrorCase.dialect !== mirror) {
      process.stdout.write(`SUITE ERROR the mirrored case for ${c.name} is still `
        + `${mirrorCase.dialect}, so nothing is being mirrored\n`);
      suiteErrors += 1;
      continue;
    }
    try {
      problem = mirrorCase.plan ? runPlanCase(mirrorCase) : runCase(mirrorCase);
    } catch (e) {
      if (!(e instanceof SuiteError)) throw e;
      process.stdout.write(`SUITE ERROR (mirrored to ${mirror}) ${e.message}\n`);
      suiteErrors += 1;
      continue;
    }
    if (problem === null) mirrored += 1;
    else {
      failures.push([c, `mirrored to ${mirror}, which must agree with ${c.dialect}: ${problem}`]);
    }
  }

  for (const [c, problem] of failures) {
    process.stdout.write(`FAIL ${c.name}  (${c.at})\n`);
    process.stdout.write(`     ${problem}\n`);
  }

  process.stdout.write(`\n${passed} passed (${mirrored} also checked against a `
    + `mirrored dialect), ${failures.length} failed, ${suiteErrors} suite errors\n`);
  return failures.length === 0 && suiteErrors === 0 ? 0 : 1;
}

process.exit(main(process.argv));
